testing/camunda-process-test-java/src/test/java/io/camunda/process/test/api/VariableAssertTest.java

```diff
@@ -518,4 +526,330 @@ void shouldFailIfProcessInstanceNotFound() {
+          .thenReturn(Collections.singletonList(COMPLEX_VARIABLE));
+      when(processInstanceEvent.getProcessInstanceKey()).thenReturn(PROCESS_INSTANCE_KEY);
+
+      // then
+      CamundaAssert.assertThat(processInstanceEvent)
+          .hasVariableSatisfies(
+              COMPLEX_VARIABLE_KEY, SimpleJsonObject.class, assertionRequirements);
+    }
+
+    @Test
+    void shouldWaitUntilVariableExists() {
+      // given
+      final Variable variableA = newVariable("a", "1");
+      final Variable variableB = newVariable("b", "2");
+
+      when(camundaDataSource.findGlobalVariablesByProcessInstanceKey(PROCESS_INSTANCE_KEY))
+          .thenReturn(Collections.singletonList(variableB))
+          .thenReturn(Arrays.asList(variableA, variableB));
+      when(processInstanceEvent.getProcessInstanceKey()).thenReturn(PROCESS_INSTANCE_KEY);
+
+      // then
+      CamundaAssert.assertThat(processInstanceEvent)
+          .hasVariableSatisfies(
+              "a", String.class, value -> Assertions.assertThat(value).isEqualTo("1"));
+
+      verify(camundaDataSource, times(2))
+          .findGlobalVariablesByProcessInstanceKey(PROCESS_INSTANCE_KEY);
+    }
+
+    @Test
+    void shouldHaveSensibleErrorMessage_assertionFailure() {
+      // given
+      when(camundaDataSource.findGlobalVariablesByProcessInstanceKey(PROCESS_INSTANCE_KEY))
+          .thenReturn(Collections.singletonList(COMPLEX_VARIABLE));
+      when(processInstanceEvent.getProcessInstanceKey()).thenReturn(PROCESS_INSTANCE_KEY);
+
+      // then
+      Assertions.assertThatThrownBy(
+              () ->
+                  CamundaAssert.assertThat(processInstanceEvent)
+                      .hasVariableSatisfies(
+                          COMPLEX_VARIABLE_KEY,
+                          SimpleJsonObject.class,
+                          Arrays.asList(
+                              result -> Assertions.assertThat(result).isNull(),
+                              result ->
+                                  Assertions.assertThat(result)
+                                      .extracting("strValue", "intValue", "boolValue")
+                                      .containsExactlyInAnyOrder("wrong", -1, false),
+                              result ->
+                                  Assertions.assertThat(result)
+                                      .extracting(
+                                          "list", InstanceOfAssertFactories.list(Object.class))
+                                      .containsExactlyInAnyOrder(
+                                          -10, "wrong", false, "truthy", Collections.emptyList()),
+                              result ->
+                                  Assertions.assertThat(result)
+                                      .extracting("nestedObject.key")
+                                      .isEqualTo("wrong_value"))))
+          .hasMessageContainingAll(
+              "Multiple Failures (4 failures)",
```

Comment:
:thought_balloon: We should provide more context in the failure message.

For example:
```
Process instance [key: 1] should have a variable 'complex' but the following requirements are not satisfied: 
```